compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignNpuDmaBdIds.cpp

```diff
@@ -268,79 +572,27 @@ LogicalResult assignNpuDmaBdIds(AMDAIE::WorkgroupOp workgroupOp) {
   // and release BD IDs when encountering the respective operations using the
   // tile BD ID generators initialized earlier.
   AMDAIE::ControlCodeOp controlCodeOp = workgroupOp.getControlCode();
-  WalkResult res = controlCodeOp->walk([&](Operation *op) {
-    if (auto npuDmaOp = dyn_cast<AMDAIE::NpuDmaCpyNdOp>(op)) {
-      if (npuDmaOp.getSource()) {
-        FailureOr<AMDAIE::BdIdOp> bdIdOp = getBdIdOp<CopyOpOperateOn::Source>(
-            rewriter, npuDmaOp, shimTileToGeneratorMap, bdIdOpToBdIdsMap);
-        if (failed(bdIdOp)) return WalkResult::interrupt();
-        rewriter.setInsertionPoint(npuDmaOp);
-        npuDmaOp = rewriter.replaceOpWithNewOp<AMDAIE::NpuDmaCpyNdOp>(
-            npuDmaOp, npuDmaOp.getResultTypes(), npuDmaOp.getConnection(),
-            npuDmaOp.getTarget(), npuDmaOp.getTargetMixedOffsets(),
-            npuDmaOp.getTargetMixedSizes(), npuDmaOp.getTargetMixedStrides(),
-            npuDmaOp.getTargetBdId(), npuDmaOp.getSource(),
-            npuDmaOp.getSourceMixedOffsets(), npuDmaOp.getSourceMixedSizes(),
-            npuDmaOp.getSourceMixedStrides(), *bdIdOp);
-      }
-      if (npuDmaOp.getTarget()) {
-        FailureOr<AMDAIE::BdIdOp> bdIdOp = getBdIdOp<CopyOpOperateOn::Target>(
-            rewriter, npuDmaOp, shimTileToGeneratorMap, bdIdOpToBdIdsMap);
-        if (failed(bdIdOp)) return WalkResult::interrupt();
-        rewriter.setInsertionPoint(npuDmaOp);
-        (void)rewriter.replaceOpWithNewOp<AMDAIE::NpuDmaCpyNdOp>(
-            npuDmaOp, npuDmaOp.getResultTypes(), npuDmaOp.getConnection(),
-            npuDmaOp.getTarget(), npuDmaOp.getTargetMixedOffsets(),
-            npuDmaOp.getTargetMixedSizes(), npuDmaOp.getTargetMixedStrides(),
-            *bdIdOp, npuDmaOp.getSource(), npuDmaOp.getSourceMixedOffsets(),
-            npuDmaOp.getSourceMixedSizes(), npuDmaOp.getSourceMixedStrides(),
-            npuDmaOp.getSourceBdId());
-      }
-      return WalkResult::advance();
-    } else if (auto npuWaitOp = dyn_cast<AMDAIE::NpuDmaWaitOp>(op)) {
-      // Release BD ID used by input DMA op.
-      for (AMDAIE::NpuDmaCpyNdOp npuDmaOp : npuWaitOp.getDmaOps()) {
-        AMDAIE::BdIdOp bdIdOp;
-        if (npuDmaOp.getSourceBdId()) {
-          bdIdOp = dyn_cast_if_present<AMDAIE::BdIdOp>(
-              npuDmaOp.getSourceBdId().getDefiningOp());
-          if (!bdIdOp) return WalkResult::advance();
-          if (failed(releaseBdId(bdIdOp, shimTileToGeneratorMap,
-                                 bdIdOpToBdIdsMap)))
-            return WalkResult::interrupt();
-        }
+  // Since a DMA op can have source and target, therefore we can have two BD IDs
+  // for any DMA op. Hence we maintain a map from DMA op to a vector of BD IDs.
+  DenseMap<AMDAIE::NpuDmaCpyNdOp, SmallVector<AMDAIE::BdIdOp>> dmaOpToBdIdMap;
+  TileDmaBatchGraph tileDmaBatchGraph = createTileDmaBatchGraph(
+      workgroupOp, controlCodeOp, bdIdAssignmentUtil.shimTileToGeneratorMap);
```

Comment:
It's a little bit strange that we need to pass the `shimTileToGeneratorMap` from `BdIdAssignmentUtil` here. I think it's better to decouple this interdependency by for example extracting `createShimTileToGeneratorMap` from `BdIdAssignmentUtil` to a standalone function, call that to initialize the map and pass it both to `createTileDmaBatchGraph` and to `BdIdAssignmentUtil`.

Reply:
Done.